ui/dashboards/src/components/Variables/VariableList.tsx

```diff
@@ -24,51 +24,197 @@ import {
   Table,
   Paper,
   TableHead,
+  Switch,
+  Typography,
+  IconButton,
 } from '@mui/material';
-import { useTemplateVariableDefinitions, useEditMode } from '../../context';
+import { VariableDefinition } from '@perses-dev/core';
+import { useImmer } from 'use-immer';
+import EyeIcon from 'mdi-material-ui/Eye';
+import PencilIcon from 'mdi-material-ui/Pencil';
+import TrashIcon from 'mdi-material-ui/TrashCan';
+
+import { useTemplateVariableDefinitions, useEditMode, useTemplateVariableActions } from '../../context';
 import { TemplateVariable } from './Variable';
+import { VariableEditForm } from './VariableEditorForm';
+
+function VariableEditor(props: {
+  variableDefinitions: VariableDefinition[];
+  onChange: (variableDefinitions: VariableDefinition[]) => void;
+  onCancel: () => void;
+}) {
+  const [variableDefinitions, setVariableDefinitions] = useImmer(props.variableDefinitions);
+  const [variableEditIdx, setVariableEditIdx] = useState<number | null>(null);
+  const isEditingVariable = variableEditIdx !== null;
+
+  const removeVariable = (index: number) => {
+    setVariableDefinitions((draft) => {
+      draft.splice(index, 1);
+    });
+  };
+
+  const addVariable = () => {
+    setVariableDefinitions((draft) => {
+      draft.push({
+        kind: 'TextVariable',
+        spec: {
+          name: 'New Variable',
+          value: '',
+        },
+      });
+    });
+  };
+
+  const toggleVariableVisibility = (index: number, visible: boolean) => {
+    setVariableDefinitions((draft) => {
+      const v = draft[index];
+      if (!v) {
+        return;
+      }
+      if (!v.spec.display) {
+        v.spec.display = {
+          hidden: false,
+        };
+      }
+      v.spec.display.hidden = visible === false;
+    });
+  };
+
+  return (
+    <Box p={4}>
+      {isEditingVariable && (
+        <>
+          <Typography variant="h3" mb={2}>
+            Edit Variable
+          </Typography>
+          <VariableEditForm
+            initialVariableDefinition={variableDefinitions[variableEditIdx] as VariableDefinition}
```

Comment:
Same idea as my previous comment: would be cool if we could figure out what's up and if we can improve types of whatever to avoid the cast.

Reply:
I believe the reason I did this is because its possible the result `variableDefinitions[variableEditIdx] === null` if `variableEditIdx` is not a valid position in the array. However, the way this value gets set is via a button that is derived from the array of definitions. Do you think its safer to save a variable in the top of the component then use that instead of `isEditingVariable`

Reply:
Yeah, sounds like if there's just a way to structure this so you can do a `null` check first somewhere (and handle that case, even if it's a `throw`), that would be the way to try and go.